File: 2022/day01/python/solution.py
#!/usr/bin/env python3
import os

def parse_input(filename):
    """Parse input into list of calorie totals per elf."""
    with open(filename) as f:
        content = f.read().strip()

    elves = []
    for group in content.split('\n\n'):
        total = sum(int(line) for line in group.split('\n') if line)
        elves.append(total)

    return elves

def part1(elves):
    """Find the Elf carrying the most Calories."""
    return max(elves)

def part2(elves):
    """Find total calories carried by top three Elves."""
    sorted_elves = sorted(elves, reverse=True)
    return sum(sorted_elves[:3])

def main():
    script_dir = os.path.dirname(os.path.abspath(__file__))
    input_file = os.path.join(script_dir, '..', 'input.txt')

    elves = parse_input(input_file)

    print('Part 1:', part1(elves))
    print('Part 2:', part2(elves))

if __name__ == '__main__':
    main()

File: 2022/day01/ruby/solution.rb
#!/usr/bin/env ruby

def parse_input(filename)
  content = File.read(filename).strip

  content.split("\n\n").map do |group|
    group.split("\n").map(&:to_i).sum
  end
end

def part1(elves)
  elves.max
end

def part2(elves)
  elves.sort.reverse.take(3).sum
end

def main
  script_dir = File.dirname(File.expand_path(__FILE__))
  input_file = File.join(script_dir, '..', 'input.txt')

  elves = parse_input(input_file)

  puts "Part 1: #{part1(elves)}"
  puts "Part 2: #{part2(elves)}"
end

main
